shims/common/src/main/scala/org/apache/gluten/config/ConfigBuilder.scala

```diff
@@ -0,0 +1,229 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.config
+
+import org.apache.spark.network.util.ByteUnit
+import org.apache.spark.network.util.JavaUtils
+
+import java.util.concurrent.TimeUnit
+import java.util.regex.Pattern;
+
+object BackendType extends Enumeration {
+  type BackendType = Value
+  val COMMON, VELOX, CLICKHOUSE = Value
+}
+
+private[gluten] case class ConfigBuilder(key: String) {
+  import ConfigHelpers._
+
+  private[config] var _doc = ""
+  private[config] var _version = ""
+  private[config] var _backend = BackendType.COMMON
+  private[config] var _public = true
+  private[config] var _alternatives = List.empty[String]
+  private[config] var _onCreate: Option[ConfigEntry[_] => Unit] = None
```

Comment:
Maybe we can add `deprecated` here, so we can deprecated some configs and warning user while user setting it.

Reply:
It should not be the case, maybe. `onCreate` occurs when the entry is created and it always happens. However, deprecation should take place when the user actively sets a deprecated key. Checking for deprecated keys from `sparkConf` during plugin initialization is more appropriate and will only issue a warning once.